### ISSUE TITLE
syntax error: found character '\t' that cannot start any token (syntax)

### ISSUE THREAD

**Issue by ReenigneArcher** [adrienverge/yamllint#507]
I have the following in a GitHub workflow. There are two tab characters in the `if` statement.

```yml
if: ${{ env.git_diff == '1	1	locale/myfile.po' }}
```

I get the following error message from yamllint.

```
[error] syntax error: found character '\t' that cannot start any token (syntax)
```

Replacing the tabs with `\t` removes the error, but then the comparison does not evaluate properly.

Adding `# yamllint disable-line` does not disable the error.

**Comment by adrienverge** [comment]
Please look for existing issues (#61, #65, #128, #311, #460, #462...) before opening new ones. Thanks!

In short:
- Your snippet is not valid YAML, it's meant to be pre-processed (by GitHub?) first. So you cannot lint it, because it's not YAML: a YAML parser fails to interpret it.
- Disable-toggles like `# yamllint disable-line` rely on the YAML parser to work.

Maybe you can try using embedding the problematic string inside a multi-line YAML string.

**Comment by ReenigneArcher** [comment]
> Your snippet is not valid YAML

Are you saying GitHub actions are not yaml? Because the code snippet I pasted works fine (it is an if condition on a job step). Only fails with linting.

Of course I searched issues here and all over the internet for days before I posted any issue. The only similar issue I saw was when someone actually used tabs instead of spaces to align the keys.

Anyway, I ended up using `sed` to replace the tab characters, which seems to work properly and satisfies the linter.

**Comment by adrienverge** [comment]
> Are you saying GitHub actions are not yaml?

It might not be valid YAML (I don't know, I lack context to understand); it depends whether `${{ … }}` depends is pre-processed before being YAML. For example, Jinja2 templates are not valid YAML (yet) because `{% … %}` doesn't mean anything in YAML (but once pre-processed it becomes valid YAML).

Please tell me if you disagree (then please give more context and/or canonize the problem to its simplest form :pray:). It could also be linked to the way our parser (PyYAML) handles tabulations.

**Comment by ReenigneArcher** [comment]
Okay, I see what you're saying now.

For additional context here is the workflow (already changed): https://github.com/LizardByte/RetroArcher/blob/44ec669d849a3c2859090c996ea774ad5ec34867/.github/workflows/localize.yml#L51

From my understanding GitHub would replace anything in these `${{ … }}` with the value it finds before it runs the action, so that makes sense... sort of. This is the only case where I've found any issue with using `${{ … }}`

This is the action I'm using to lint the yaml files. https://github.com/ibiqlik/action-yamllint

Although I asked the question here because I also have the yamllint plugin in PyCharm highlighting the same error.

**Comment by perlpunk** [comment]
You have to quote the string if it contains literal tabs.
They are valid in plain strings in YAML, however, PyYAML (which yamllint is based on) diverges from the spec here and forbids them.
suggestion:
```
if: "${{ env.git_diff == '1	1	locale/myfile.po' }}"
```
or
```
if: |
  ${{ env.git_diff == '1	1	locale/myfile.po' }}
```

**Comment by adrienverge** [comment]
OK thanks for the details!

I don't know whether or not GitHub replaces `${{ … }}` *before* or *after* parsing YAML, it would need some research.

But I also found that PyYAML (the parser behind yamllint) can fail on `\t` characters:
> rules for tabs in YAML are confusing. We are close, but not there yet. Perhaps both the spec and the parser should be fixed. Anyway, the best rule for tabs in YAML is to not use them at all.
> -- https://pyyaml.org/wiki/PyYAMLDocumentation#deviations-from-the-specification

**Comment by perlpunk** [comment]
Regarding the github action expression syntax: The expressions are replaced (similar to Ansible) _after_ loading the YAML, so github workflow files should be valid YAML and you can use yamllint on it.

**Comment by ReenigneArcher** [comment]
https://github.com/yaml/pyyaml/issues/450

This appears to be the same. I'm not 100% sure since it mentions a trailing tab.
